dht_test.go

```diff
@@ -1107,6 +1106,55 @@ func TestBadProtoMessages(t *testing.T) {
 	}
 }
 
+func TestAtomicPut(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	d := setupDHT(ctx, t, false)
+	d.Validator = testValidator{}
+
+	// fnc to put record
+	key := "testkey"
+	putRecord := func(value []byte) error {
+		rec := record.MakePutRecord(key, value)
+		pmes := pb.NewMessage(pb.Message_PUT_VALUE, rec.Key, 0)
+		pmes.Record = rec
+		_, err := d.handlePutValue(ctx, "testpeer", pmes)
+		return err
+	}
+
+	// put a valid record
+	if err := putRecord([]byte("valid")); err != nil {
+		t.Fatal("should not have errored on valid record")
+	}
+
+	// start 30 threads to put `old` & `new` records
+	var wg sync.WaitGroup
+	for i := 0; i < 30; i++ {
+		wg.Add(1)
+		go func(i int) {
+			defer wg.Done()
+			if rand.Intn(2) == 0 {
+				if err := putRecord([]byte("newer")); err != nil {
```

Comment:
This will fail if we've already "put" the newer value. What we need to do is race putting the old value with the new value and then check to make sure we only see the new one (likely constructing a new DHT each time).

Reply:
@Stebalien  Yeah, I understand that. The test was still WIP, but I wanted to see the result of the build. A few questions for you :

1) What do you mean by constructing a new DHT each time ?
2) Are you able to run the dht travis build locally ? If yes, how do you do it ?
3) Why do most DHT keys start with a "/"

Reply:
> What do you mean by constructing a new DHT each time ?

`setupDHT()`. Basically, once the "newer" put succeeds, you'll have to create a new DHT to re-try the race.

> Are you able to run the dht travis build locally ? If yes, how do you do it ?

`go test ./...`

> Why do most DHT keys start with a "/"

They're namespaced. DHT keys are usually `/$namespace/$key` where `$namespace` is currently either `pk` (public key) or `ipns`.